### PR TITLE
Add basic pipeline structure

### PR DESCRIPTION
**Pull request by mradul2**
Basic pipeline for loading data, training and validating the model along with logging and saving on WandB. Some files are to be modified and fixed. Please check and suggest changes.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,6 +1,7 @@
 # AGV DL Segmentation Pipeline
```
Comment (karan-uppal3) [inline README.md]: Do complete the readme with the directory structure so its easier to go through the next time

File: dataloader/cityscapes_cutmix.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    self.transform = transforms
+    self.num_mix = num_mix
+    self.beta = beta
+    self.prob = prob
+
+  def __getitem__(self, index):
+    image = (self.data[index]).copy()
+    label = (self.label[index]).copy()
+
+    for _ in range(self.num_mix):
```
Comment (karan-uppal3) [inline dataloader/cityscapes_cutmix.py]: This would apply Cutix to each individual image and probably lead to more computations. There's a pseudo-code present in the CutMix paper that performs the operation for a batch of images in the training loop itself. Do check that out

A better way to go about this would be to have a bool parameter whether you want to apply CutMix or not and pass that bool during training. In the training loop, the batch can be passed to a function that applies CutMix depending on the value of the bool parameter.

File: run.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
+# from utils.trainer import evaluate
+# import train, test
+# from utils.config import process_config
```
Comment (karan-uppal3) [inline run.py]: Some functions like `process_config` haven't been imported but have been used below

File: train.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+                                          weight_decay=self.config.weight_decay)
+
+        self.scheduler = torch.optim.lr_scheduler.ExponentialLR(self.optimizer,
+                                                                gamma=self.config.gamma)
+
+        self.valid_dataloader = self.dataloader.valid_loader
+        self.train_dataloader = self.dataloader.train_loader
+
+        self.current_epoch = 1
+        self.min_valid_loss = 10.0
```
Comment (karan-uppal3) [inline train.py]: Better to keep it as np.inf

File: train.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+        train_accuracy, valid_accuracy, train_iou, valid_iou, valid_results = final_metrics(self.config,
+                                                                         self.model,
+                                                                         self.train_loader,
+                                                                         self.valid_loader,
+                                                                         self.device)
+
+        if self.config.wandb: 
+            print("Saving Experiment summary on WandB")
+            save_model_wandb(self.config.checkpoint_path)
+            save_model_wandb
```
Comment (karan-uppal3) [inline train.py]: do remove the line 105

File: utils/tester.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
+from metric import iou_accu
+import numpy as np
+
+# These are utility functions, we need not to initialize any dataloader or model here (ENet or Cityscapes)
+# Just pass those in these functions wherever these are called (test.py or train.py) ~mradul
+
+def final_metrics(config, model, train_loader, valid_loader, device):
```
Comment (karan-uppal3) [inline utils/tester.py:7]: the function `final_metrics` in `utils/tester.py` and `validate` in `utils/trainer.py` are essentially performing the same operation. So instead have only one function which will take as arguments the dataloader, model, config and device and output the corresponding loss and accuracy. So during the validation step of training, you can even log the accuracy and it would shorten this code as well

File: utils/wandb_utils.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+    Initialize project on Weights & Biases
+    Args:
+        model (Torch Model): Model for Training
+        args (TrainOptions,optional): TrainOptions class (refer options/train_options.py). Defaults to None.
+    """
+
+    print("Wandb api key provided...")
+    wandb.login(key=config.wandb_id)
+
+    wandb.init(
```
Comment (karan-uppal3) [inline utils/wandb_utils.py:20]: Log the config file using the config argument of `wandb.init`

File: dataloader/cityscapes_cutmix.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+            continue
+
+        lam = np.random.beta(self.beta, self.beta)
+        rand_index = random.choice(range(len(self)))
+
+
+        image2 = self.data[rand_index]
+        label2 = self.label[rand_index]
+
+        bbx1, bby1, bbx2, bby2 = rand_bbox(image.shape, lam)
```
Comment (karan-uppal3) [inline dataloader/cityscapes_cutmix.py]: Function `rand_bbox` hasn't been imported above

File: test.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+        self.current_epoch = 1
+
+        self.device = default_device()
+
+        self.model = self.model.to(self.device)
+        self.loss = self.loss.to(self.device)
+
+    def forward(self):
+      tester(epochs=self.epochs, model=self.model, train_loader=self.train_loader, val_loader=self.valid_loader, criterion=self.criterion, optimizer=self.optimizer, scheduler=self.scheduler, checkpoint_path, best_model_path)
```
Comment (karan-uppal3) [inline test.py:40]: For testing purpose, you don't require things like optimizer and learning rate schedulers

File: utils/wandb_utils.py [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+        val_acc (float): Validation Accuracy
+        epoch (int): Epoch Number
+    """
+
+    wandb.log({
+        'Loss/Training': train_loss,
+        'Loss/Validation': val_loss,
+    }, step=epoch)
+
+def wandb_save_summary(valid_accuracy,
```
Comment (karan-uppal3) [inline utils/wandb_utils.py:45]: Do read about artifacts on wandb as well, for saving the best model and the last checkpoint model to wandb, and add this feature as well if possible

**Review by karan-uppal3 — COMMENTED** [review]
Overall pretty great work, nice structure of the repository. A few quirks are present (CutMix augmentation) and some things are yet to be done (running on the test set), do work on them and all the issues that I've pointed out as well.
